File: Overload/Complex.py

class Complex:

    def __init__(self, real, complex):
        self.real = real
        self.complex = complex

    def __str__(self):
        if self.real == 0 and self.complex == 0:
            return "0"
        elif self.real == 0:
            return str(self.complex) + "j"
        elif self.complex == 0:
            return str(self.real)
        elif self.complex > 0:
            if self.complex == 1:
                return str(self.real) + "+j"
            else:
                return str(self.real) + "+" + str(self.complex) + "j"
        else:
            if self.complex == -1:
                return str(self.real) + "-j"
            else:
                return str(self.real) + str(self.complex) + "j"

    
    def __add__(self, other):
        if isinstance(other, Complex):
            return Complex(self.real + other.real, self.complex + other.complex)
        elif isinstance(other, int) or isinstance(other, float):
            return self + Complex(other, 0)
        else: # something is wrong
            raise TypeError

    def __sub__(self, other):
        if isinstance(other, Complex):
            return Complex(self.real - other.real, self.complex - other.complex)
        elif isinstance(other, int) or isinstance(other, float):
            return self - Complex(other, 0)
        else: # something is wrong
            raise TypeError

    def __neg__(self, other):
        return Complex(-self.real, -self.complex)


    def __mul__(self, other):
        if isinstance(other, Complex):
            real = self.real * other.real - self.complex * other.complex
            complex = self.real * other.complex + self.complex * other.real
            return Complex(real, complex)
        elif isinstance(other, int) or isinstance(other, float):
            return self * Complex(other, 0)
        else:
            raise TypeError

    def __truediv__(self, other):
        if isinstance(other, Complex):
            result = self * Complex(other.real, -other.complex)
            ratio = other.real * other.real + other.complex * other.complex
            result.real /= ratio
            result.complex/= ratio
            return result
        elif isinstance(other, int) or isinstance(other, float):
            return self / Complex(other, 0)
        else: 
            raise TypeError


    def __radd__(self, other):
        return self + other

    def __rsub__(self, other):
        return Complex(0, 0) + other - self

    def __rmul__(self, other):
        return self * other

    def __rtruediv__(self, other):
        return Complex(1, 0) * other / self


#  test program
def main():
     
    d = float (input ("Enter any value with decimal point: "))
    
    
    i= int (input ("Enter the real part of a: " ))
    
    j = int (input ("Enter the complex part of a: "))

    x = int(input ("Enter the real part of b: "))
    
    y = int (input("Enter the complex part of b: "))
    
    a = Complex (i, j)
    b = Complex (x, y)

    
    

    print("a = %s" % (a))
    print("b = %s" % (b))

    # operators

    # two complex
    print("a + b = %s" % (a + b))
    print("a - b = %s" % (a - b))
    print("a * b = %s" % (a * b))
    print("a / b = %s" % (a / b))

    # complex with double
    print("a + d = %s" % (a + d))
    print("a - d = %s" % (a - d))
    print("a * d = %s" % (a * d))
    print("a / d = %s" % (a / d))

    # double with complex
    print("d + b = %s" % (-d + b))
    print("d - b = %s" % (d - b))
    print("d * b = %s" % (d * b))
    print("d / b = %s" % (d / b))

main()

File: Overload/Complex.java
public class Complex {

  private final double real;
  private final double complex;

  public Complex(double real, double complex) {
    this.real = real;
    this.complex = complex;
  }

  public double getReal() {
    return real;
  }

  public double getComplex() {
    return complex;
  }

  public Complex add(Complex other) {
    return new Complex(this.real + other.real, this.complex + other.complex);
  }

  public Complex subtract(Complex other) {
    return new Complex(this.real - other.real, this.complex - other.complex);
  }

  public Complex multiply(Complex other) {
    double real = this.real * other.real - this.complex * other.complex;
    double complex = this.real * other.complex + this.complex * other.real;
    return new Complex(real, complex);
  }

  public Complex divide(Complex other) {
    Complex result = this.multiply(new Complex(other.real, - other.complex));
    double val = other.real * other.real + other.complex * other.complex;
    return new Complex(result.real / val, result.complex / val);
  }

  public Complex add(double other) {
    return this.add(new Complex(other, 0));
  }

  public Complex subtract(double other) {
    return this.subtract(new Complex(other, 0));
  }

  public Complex multiply(double other) {
    return this.multiply(new Complex(other, 0));
  }

  public Complex divide(double other) {
    return this.divide(new Complex(other, 0));
  }

  @Override
  public String toString() {
    Complex num= this;
    if (num.real == 0 && num.complex == 0) {
      return "num";
    } else if (num.real == 0) {
      return "" + num.complex + "j";
    } else if (num.complex == 0) {
      return "" + num.real;
    } else if (num.complex > 0) {
      if (num.complex == 1) {
        return "" + num.real + "+j";
      } else {
        return "" + num.real + "+" +  num.complex + "j";
      }
    } else {
      if (num.complex == -1) {
        return "" + num.real + "-j";
      } else {
        return "" + num.real + num.complex + "j";
      }
    }
  }
}